### PR TITLE
Add FirstData E4 Support for Multiple Mastercard String Formats

### PR DESCRIPTION
**Pull request by jcbantuelle**
The response back from FirstData uses the string 'mastercard'. When using the FirstData Response to reconstruct the credit card info, it needs this key in the BRANDS constant to convert to a valid card.

### CODE REVIEW

File: test/unit/gateways/firstdata_e4_test.rb [hunk 1]
```diff
@@ -32,7 +36,7 @@ def test_successful_purchase
     assert response.test?
     assert_equal 'Transaction Normal - Approved', response.message
 
-    ExactGateway::SENSITIVE_FIELDS.each{|f| assert !response.params.has_key?(f.to_s)}
```
Comment (j-mutter) [inline test/unit/gateways/firstdata_e4_test.rb]: Oof, nice catch

File: lib/active_merchant/billing/gateways/firstdata_e4.rb [hunk 2]
```diff
@@ -263,7 +264,7 @@ def commit(action, request, credit_card = nil)
 
         Response.new(successful?(response), message_from(response), response,
           :test => test?,
-          :authorization => response_authorization(action, response, credit_card),
+          :authorization => successful?(response) ? response_authorization(action, response, credit_card) : nil,
```
Comment (j-mutter) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: I'm a little concerned about this change. When `#response_authorization` falls through to `#authorization_from`,with what appears to be an unsuccessful transaction, it [returns an empty string](https://github.com/jcbantuelle/active_merchant/blob/master/lib/active_merchant/billing/gateways/firstdata_e4.rb#L292-L294) not `nil`, so I'm worried that this change in behaviour may break some integrations that expect to always have a string here.

Comment (ntalbott) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: @j-mutter but AM makes no guarantees around `nil` vs. `""` for `Response#authorization`, so I don't think we need to maintain this behavior. It's certainly not maintained on other adapters.

Comment (j-mutter) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: Yeah I know, just might be a nasty surprise for someone

Comment (ntalbott) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: That'll teach 'em! :wink:

I don't really care either way in this specific case, just want to be cautious in general about making anyone think we support their broken calling code. :fearful:

Comment (jcbantuelle) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: I've updated the fallback value to be an empty string instead of nil for consistency.

**Comment by jcbantuelle** [comment]
Just checking in, is there anything else you need from me to get this merged in?

**Comment by jcbantuelle** [comment]
ping @ntalbott @j-mutter

File: lib/active_merchant/billing/gateways/firstdata_e4.rb [hunk 3]
```diff
@@ -263,7 +264,7 @@ def commit(action, request, credit_card = nil)
 
         Response.new(successful?(response), message_from(response), response,
           :test => test?,
-          :authorization => response_authorization(action, response, credit_card),
+          :authorization => successful?(response) ? response_authorization(action, response, credit_card) : '',
```
Comment (j-mutter) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: Can you add an assertion for this change to your unsuccessful test?

Comment (jcbantuelle) [inline lib/active_merchant/billing/gateways/firstdata_e4.rb]: @j-mutter Thanks for the feedback! I've added an assertion for this, please let me know if you need any further additions.

**Comment by jcbantuelle** [comment]
ping @ntalbott @j-mutter

**Comment by ntalbott** [comment]
This PR confuses me; does AM ever use `:mastercard` as the card type? That seems like a bug, if so.

**Comment by jcbantuelle** [comment]
@ntalbott The issue isn't with AM, it's with how FirstData E4 tokenizes credit card information. If you look at https://github.com/Shopify/active_merchant/blob/master/lib/active_merchant/billing/gateways/firstdata_e4.rb#L200, you'll see that it calls credit_card.brand. When using the FirstData token response from their server, that brand value is 'mastercard'. The card_type function at https://github.com/Shopify/active_merchant/blob/master/lib/active_merchant/billing/gateways/firstdata_e4.rb#L262-L264 uses the BRANDS constant to convert the brand value, and without :mastercard in BRANDS, this conversion fails.

**Comment by ntalbott** [comment]
@jcbantuelle OK, got it. My fear is that we're overloading `supported_cardtypes` with a value that will confuse any code using it as metadata. Can we do this without leaking the `:mastercard` key outside this adapter?

**Comment by jcbantuelle** [comment]
@ntalbott Thank you for the feedback, that makes a lot of sense. I've pulled the mastercard value out into a separate hash that is used exclusively in the card_type function to keep this behavior encapsulated in the FirstData E4 adapter. Please let me know if you need any further changes.

**Comment by jcbantuelle** [comment]
ping @ntalbott @j-mutter

**Comment by j-mutter** [comment]
Looks fine to me. @ntalbott @wvanbergen @girasquid anyone want to give this a second +1 before I merge it?

**Comment by ntalbott** [comment]
:shipit:

**Pull request closed without merging**